feat: tool to validate state witness

Longarithm opened the pull request:
Completes the lifetime of state witness saved to disk:

Save latest state witnesses for given selection:
```
neard --unsafe-fast-startup view-state --readwrite state-witness latest --shard-id 4 --height 118772867 --binary > witness.txt
```

To validate it:
```
RUST_LOG=debug neard --unsafe-fast-startup view-state --readwrite state-witness validate --input-file witness.txt
```

The huge diff is explained by the need to move lots of state witness validation code and metrics from near-client to near-chain, and it actually belong there. Probably quick look is enough to check that I didn't break it. 

The only new logic is `ValidateWitnessCmd` which just calls existing function for shadow validation. I checked manually that with `RUST_LOG=debug` it shows error on invalid witness and success on the valid one.

jancionear commented on tools/state-viewer/src/latest_witnesses.rs:
```diff
@@ -57,3 +82,53 @@ impl LatestWitnessesCmd {
+                runtime_adapter.as_ref(),
+            )
+            .unwrap();
+        // Not optimal, but needed to wait for potential `validate_state_witness` failure.
+        loop {
+            std::thread::sleep(core::time::Duration::from_secs(60));
+        }
+    }
```
:cry: 

Maybe you could add a `ProcessingDoneTracker` argument to `shadow_validate_state_witness`, like it's done in `process_chunk_state_witness`/`start_validating_chunk`

https://github.com/near/nearcore/blob/62bc785f2a94cae7ecd72ec61363bb57793c458b/chain/client/src/stateless_validation/chunk_validator/mod.rs#L115

jancionear commented on tools/state-viewer/src/latest_witnesses.rs:
```diff
@@ -57,3 +82,53 @@ impl LatestWitnessesCmd {
+
+impl ValidateWitnessCmd {
+    pub(crate) fn run(&self, home_dir: &Path, near_config: NearConfig, store: Store) {
+        let encoded_witness: Vec<u8> =
+            serde_json::from_reader(BufReader::new(std::fs::File::open(&self.input_file).unwrap()))
```
I wonder if the serialization format is user-friendly.
The `--binary` just debug-prints a `Vec<u8>`, at the time I thought that it would make it easy to paste into code or something, but now on second thought I think it might be annoying. Pasting multiple megabytes of text into VSCode sounds like a bad idea.
Apart from that the command can print out multiple witnesses in multiple lines, along with some other information.

Maybe it would be better if the `--binary` flag outputted a serialized `Vec<ChunkStateWitness>` as raw non-text bytes? That could be saved to a file and read directly, without treating the array as json (although that's pretty clever!).

@Longarithm you actually used the tool in the wild to debug stuff, do you think the current format is usable? WDYT?

pugachAG commented on chain/chain/src/metrics.rs:
```diff
@@ -271,3 +273,172 @@ pub static SAVED_LATEST_WITNESSES_SIZE: Lazy<IntGauge> = Lazy::new(|| {
     )
     .unwrap()
 });
+
```
let's move this to a separate module to avoid having large files, something like `chain/chain/src/stateless_validation/metrics.rs`

pugachAG commented on chain/chain/src/stateless_validation/mod.rs:
```diff
@@ -1,2 +1,3 @@
 pub(crate) mod chunk_endorsement;
 pub(crate) mod state_transition_data;
+pub mod state_witness;
```
it doesn't seem like a descriptive module name, I find it rather confusing 🙃 
I suggest changing it to `chunk_validation`

pugachAG reviewed (changes requested):
overall LGTM

Longarithm commented:
Updates:
* Added ProcessingDoneTracker to shadow validation, so tool exits immediately on result;
* Moved metrics to separate file;
* Leaving only two state witness latest cmd mode: either pretty-print or screen or save raw [u8] to file.

jancionear commented on tools/state-viewer/src/latest_witnesses.rs:
```diff
@@ -38,22 +66,86 @@ impl LatestWitnessesCmd {
+                    println!("");
+                }
+                LatestWitnessesMode::Binary { ref output_dir } => {
+                    let file_name = format!(
+                        "witness_{}_{}_{}.bin",
+                        witness.chunk_header.height_created(),
+                        witness.chunk_header.shard_id(),
+                        witness.epoch_id.0
+                    );
```
Theoretically there could be multiple saved witnesses with the same `(height, shard_id, epoch_id)`. Let's maybe add the index `i` as well?

jancionear commented on tools/state-viewer/src/latest_witnesses.rs:
```diff
@@ -38,22 +66,86 @@ impl LatestWitnessesCmd {
+                Some(processing_done_tracker),
+            )
+            .unwrap();
+        waiter.wait();
```
:heart:

A message like "OK" would be nice, just to make it clear that the validation passed.

Longarithm replied:
Hmm, the waiter doesn't return anything by default... I think `ProcessingDoneTracker/Waiter` must become a `OnceCell<bool>` to propagate result properly, otherwise it can't be retrieved.
Can I make it a separate issue and ask caller to use `RUST_LOG=debug` for now? :)

jancionear replied:
Oh you're right, for a moment I thought that it either fails at `unwrap()` or succeeds, but it doesn't work like that.

wacban commented on tools/state-viewer/src/latest_witnesses.rs:
```diff
@@ -1,12 +1,36 @@
-#[derive(Parser)]
+#[derive(clap::Subcommand)]
+pub enum StateWitnessCmd {
+    /// Prints latest state witnesses saved to DB.
+    Latest(LatestWitnessesCmd),
```
Not sure if there is any existing convention to call it that but to me it's very confusing. Should that be called something like Dump/Print/Get/Write/Show WitnessCmd?

wacban commented on tools/state-viewer/src/latest_witnesses.rs:
```diff
@@ -1,12 +1,36 @@
+use nearcore::NightshadeRuntimeExt;
 
-#[derive(Parser)]
+#[derive(clap::Subcommand)]
+pub enum StateWitnessCmd {
```
Can you add a short comment on how to use both commands to first dump the witness to a file and then validate it? Also it would be nice to make it so that it's visible in the --help of this subcommand.

Longarithm replied:
Help for subcommands is taken from comment for the parent command, so I put the comment into `StateViewerSubCommand::StateWitness`.

wacban commented on tools/state-viewer/src/latest_witnesses.rs:
```diff
@@ -20,13 +44,17 @@ pub struct LatestWitnessesCmd {
+enum LatestWitnessesMode {
+    /// Pretty-print on screen using the "{:#?}" formatting.
+    Pretty,
+    /// Saves the raw &[u8] of each witness to the given directory.
+    Binary { output_dir: PathBuf },
```
This is totally sufficient but if you fancy you can make it a bit quicker by adding some sort of selector to only pick the witnesses at desired height and shard. Totally optional as this PR is large enough already :)

Longarithm replied:
`DumpWitnessesCmd` (Latest) already supports filtering by height, shard and epoch. Or you mean giving height and shard to `Validate` command? But it supports only one witness anyway.

wacban replied:
Ah, sorry, I missed that. It's fine as is, thanks!

wacban reviewed (approved):
LGTM but I only reviewed the subcommand assuming the rest is just a refactor. Let me know if there is any other part that you'd like reviewed.

wacban commented:
This is going to be a fantastic debugging tool, thanks for looking into it!

pugachAG reviewed (approved):
Thanks for addressing my comments!

Longarithm commented:
How I made sure that the PR doesn't break stuff:

```
git switch master
sed -n '234,706 p' chain/client/src/stateless_validation/chunk_validator/mod.rs > /tmp/base.rs
git switch validate-witness
diff /tmp/base.rs chain/chain/src/stateless_validation/chunk_validation.rs > diff.txt
```

[diff.txt](https://github.com/user-attachments/files/15540301/diff.txt)

Longarithm commented:
merging because we need it at #11469
